MPTokenIssuanceID New Construct and Serialization Format (#161)
XRPLF/XRPL-Standards: XLS-0033d-multi-purpose-tokens/README.md

##### 1.2.1.1.1. **`MPTokenIssuance`** Ledger Identifier

The key of a `MPTokenIssuance` object, is the result of SHA512-Half of the following values, concatenated in order:

* The MPTokenIssuance space key (0x007E).
* The transaction sequence number.
* The AccountID of the issuer.

The ID of a `MPTokenIssuance` object, a.k.a. `MPTokenIssuanceID`, is a 192-bit integer, concatenated in order:

* The transaction sequence number.
* The AccountID of the issuer.

```
┌──────────────────────────┐┌──────────────────────────┐
│                          ││                          │
│      Sequence            ││     Issuer AccountID     │
│      (32 bits)           ││        (160 bits)        │
│                          ││                          │
└──────────────────────────┘└──────────────────────────┘
```

> **Note: The `MPTokenIssuanceID` is what the user inputs for the interface. Internally, the ledger splits the `MPTokenIssuanceID` into two components: `sequence` and `issuer` address.**

Reviewer: > **Note: The `MPTokenIssuanceID` is what the user inputs for the interface.

Let's clarify this a bit more. E.g., which interface are we referring to here? I think there's a difference between API "interfaces" vs. the contents of a signed transaction, and on this point these proposed changes are a bit ambiguous.

The original intent of the `MPTokenNode` object is that it would be a sort of "directory" (i.e., an index) that stores a list of `MPTokenID` values (each 32 bytes) that exist for a single `MPTokenIssuance`. This would allow rippled to contain an RPC endpoint that could return a paged collection of `MPToken` objects for a given issuance, or somethign similar like an RPC called `mpt_holder_balances`. In theory, this could also enable rippled to operate a sort of "clean-up" operation that could remove dangling MPTokens that still live on a ledger after a corresponding `MPTokenIssuance` has been deleted (and thus return ledger reserves back to token holders).

> #### 2.2.3.1 Should We Have `MPTokenNode` Directories?

Reviewer: Why is this section being removed?

Author: Oops my bad. Just added it back.

|        FIELD NAME | SIZE (BITS) | SIZE (BYTES) | NOTE                                                                                                                                                       |
|------------------:|:-----------:|:------------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
|   LedgerEntryType |     16      |      2       |                                                                                                                                                            |

> | MPTokenIssuanceID |     192     |      32      |                                                                                                                                                            |

Reviewer: ```suggestion
| MPTokenIssuanceID |     192     |      24      |                                                                                                                                                            |
```

|             Flags |     32      |      4       |                                                                                                                                                            |
|     PreviousTxnID |     256     |      32      |                                                                                                                                                            |
| PreviousTxnLgrSeq |     32      |      4       |                                                                                                                                                            |
|  Field+Type Codes |     112     |      14      | For every field, there is a `FieldCode` and a `TypeCode`, taking 2 bytes in total (e.g., if there are 4 fields, we'll use 8 bytes). Here we have 7 fields. |
|               --- |     --      |     ---      |                                                                                                                                                            |

> |             TOTAL |     768     |     104      |                                                                                                                                                            |

Reviewer: ```suggestion
|             TOTAL |     768     |     96      |                                                                                                                                                            |
```

| Bytes for holding 10 Tokens |         1         |       1,450       |            2             |          3,260          |             2.2x              |
| Bytes for holding 32 Tokens |         2         |       5,556       |            4             |         12,264          |             2.2x              |
| Bytes for holding 64 Tokens |         3         |      13,070       |            6             |         28,444          |             2.2x              |

### 2.3.2. `STAmount` serialization
Referenced from https://gist.github.com/sappenin/2c923bb249d4e9dd153e2e5f32f96d92 with some modifications:

#### Binary Encoding

> To support this idea, we first need a way to leverage the current [STAmount](https://xrpl.org/serialization.html#amount-fields) binary encoding. To accomplish this, we notice that for XRP amounts, the maximum amount of XRP (10^17 drops) only requires 57 bits. However, in the current `XRP` amount encoding, there are 62 bits available, so we can repurpose one to indicate if an amount is indeed a MPT or not.

Reviewer: ```suggestion
To support this idea, we first need a way to leverage the current [STAmount](https://xrpl.org/serialization.html#amount-fields) binary encoding. To accomplish this, we notice that for XRP amounts, the maximum amount of XRP (10^17 drops) only requires 57 bits. However, in the current `XRP` STAmount encoding, there are 62 bits available. So, so we can repurpose one of these bits to indicate if an amount is indeed a MPT or not (and still have 4 bits left over for future use, if needed). 
```

> The rules for reading the binary amount fields would be backward compatible, as follows:

Reviewer: ```suggestion
This enables MPT amounts to be represented in the current `STAmount` binary encoding.  he rules for reading the binary amount fields would be backward compatible, as follows:
```

1. Parse off the Field ID with a type_code (`STI_AMOUNT`). This indicates the following bytes are an `STAmount`.

> 2. Inspect the next bit. If `1`, then this is **not** a MPT or XRP (instead this is a regular IOU token amount). However, if the first bit is `0`, then continue.

Reviewer: ```suggestion
2. Inspect the next bit. If its value is `1`, then continue to the next step. If not, then this `STAmount` does **not** represent an MPT nor XRP (instead this is a regular IOU token amount, and can be parsed according to existing rules for those amounts). 
```

3. Ignore (for now) the 2nd bit (this is the sign-bit, and is always 1 for both XRP and MPT).

> 4. Inspect the 3rd bit. If `0`, then parse as an XRP value per usual. However, if `1`, then parse the bytes as a MPT.

Reviewer: ```suggestion
4. Inspect the 3rd bit. If `0`, then parse as an XRP value per usual. However, if `1`, then parse the remaining `STAmount` bytes as an MPT.
```

│ │                 ││                        ││            ││                integer drops                │ │
│ │  "Not XRP" bit  ││  Sign bit (always 1;   ││"IsMPT" bit ││                  (61 bits)                  │ │
│ │0=XRP/MPT; 1=IOU ││       positive)        ││0=XRP; 1=MPT││                                             │ │
│ └─────────────────┘└────────────────────────┘└────────────┘└─────────────────────────────────────────────┘ │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

#### Encoding for MPT Values (backward compatible)

This encoding focuses on the first 3 bytes of a MPT:

```
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ ┌─────────────────┐┌────────────────────────┐┌────────────┐┌─────────────────────────────────────────────┐ │
│ │        0        ││           1            ││     1      ││                                             │ │
│ │                 ││                        ││            ││             Remaining MPT Bytes             │ │

> │ │  "Not XRP" bit  ││  Sign bit (always 1;   ││"IsMPT" bit ││                 (389 bits)                  │ │

Reviewer: ```suggestion
│ │  "Not XRP" bit  ││  Sign bit (always 1;   ││"IsMPT" bit ││                 (256 bits)                  │ │
```

│ ┌─────────────────┐┌──────────────────┐┌──────────────────────┐┌──────────────────────────┐┌──────────────────────────┐│
│ │    [0][1][1]    ││                  ││                      ││                          ││                          ││
│ │                 ││     Reserved     ││   MPT Amount Value   ││      Sequence            ││     Issuer AcountID      ││
│ │   IsMPT=true    ││     (5 bits)     ││      (64 bits)       ││      (32 bits)           ││        (160 bits)        ││
│ │                 ││                  ││                      ││                          ││                          ││
│ └─────────────────┘└──────────────────┘└──────────────────────┘└──────────────────────────┘└──────────────────────────┘│
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Note: MPT introduces an extra leading byte in front of the MPT value. However, despite the MPT value being 64-bit, only 63 bits can be used. This limitation arises because internally, rippled needs to convert the value to `int64`, which has a smaller positive range compared to `uint64`.

> *Question*: Could the ledger perform arithmetics in `uint64`? If not, we would need to decrease the range of the MPT value.

Reviewer: ```suggestion
```

I think this question is answered. We should transition this to an FAQ answer instead.

> The reasons for having a construct `MPTokenIssuanceID` by concatenating `sequence` and `issuer` are:

Reviewer: This should be formulated into the FAQ. Maybe let's remove this from this PR, and then make a different PR to update the FAQ?

Author: Sounds good!

This enables MPT amounts to be represented in the current `STAmount` binary encoding.  he rules for reading the binary amount fields would be backward compatible, as follows:

1. Parse off the Field ID with a type_code (`STI_AMOUNT`). This indicates the following bytes are an `STAmount`.
2. Inspect the next bit. If its value is `1`, then continue to the next step. If not, then this `STAmount` does **not** represent an MPT nor XRP (instead this is a regular IOU token amount, and can be parsed according to existing rules for those amounts). 
3. Ignore (for now) the 2nd bit (this is the sign-bit, and is always 1 for both XRP and MPT).
4. Inspect the 3rd bit. If `0`, then parse as an XRP value per usual. However, if `1`, then parse the remaining `STAmount` bytes as an MPT.

#### Encoding for XRP Values (backward compatible)

> ```

Reviewer: After talking to @shawnxie999 we agreed to adjust this section to be this:

```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ ┌──────────────────┐ ┌──────────────────────┐ ┌──────────────────┐ ┌──────────────────┐ ┌───────────────────────────────────────────┐ │
│ |        0         | |          1           | |        0         | |                  | |                                           | │
│ |                  | |                      | |                  | |     Reserved     | |                XRP Amount                 | │
│ |  "Not XRP" bit   | | Sign bit (always 1;  | |   "IsMPT" bit    | |     (4 bits)     | |                 (57 bits)                 | │
│ | 0=XRP/MPT; 1=IOU | |      positive)       | |   0=XRP; 1=MPT   | |                  | |                                           | │
│ └──────────────────┘ └──────────────────────┘ └──────────────────┘ └──────────────────┘ └───────────────────────────────────────────┘ │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Once that change is made, I'll merge this PR.

Reviewer: LGTM

┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ ┌─────────────────┐┌────────────────────────┐┌────────────┐┌─────────────────────────────────────────────┐ │
│ │        0        ││           1            ││     0      ││                                             │ │
│ │                 ││                        ││            ││                integer drops                │ │
│ │  "Not XRP" bit  ││  Sign bit (always 1;   ││"IsMPT" bit ││                  (61 bits)                  │ │
│ │0=XRP/MPT; 1=IOU ││       positive)        ││0=XRP; 1=MPT││                                             │ │
│ └─────────────────┘└────────────────────────┘└────────────┘└─────────────────────────────────────────────┘ │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

#### Encoding for MPT Values (backward compatible)

This encoding focuses on the first 3 bytes of a MPT:

```

> ┌────────────────────────────────────────────────────────────────────────────────────────────────────────────┐

Reviewer: After talking to @shawnxie999 we agreed to adjust this section to be this:

```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ ┌──────────────────┐ ┌──────────────────────┐ ┌──────────────────┐ ┌──────────────────┐ ┌───────────────────────────────────────────┐ │
│ |        0         | |          1           | |        0         | |                  | |                                           | │
│ |                  | |                      | |                  | |     Reserved     | |                XRP Amount                 | │
│ |  "Not XRP" bit   | | Sign bit (always 1;  | |   "IsMPT" bit    | |     (4 bits)     | |                 (57 bits)                 | │
│ | 0=XRP/MPT; 1=IOU | |      positive)       | |   0=XRP; 1=MPT   | |                  | |                                           | │
│ └──────────────────┘ └──────────────────────┘ └──────────────────┘ └──────────────────┘ └───────────────────────────────────────────┘ │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Once that change is made, I'll merge this PR.
